### ISSUE TITLE
Integration occasionally drops connectivity

### ISSUE THREAD

**Issue by txwindsurfer** [sebr/bhyve-home-assistant#28]
Running 2.0.1 after trying the alpha version.  Occasionally (1 per week) getting a websocket error which disconnects from the bhyve service.  After this error, the component runs services but they do not function.  No error is produced. 
 by the component so apparently the component believes the websocket is still functioning correctly.  The debug log and dump show no errors.  A reboot re-establishes the websocket and all is fine until next websocket error happens.  After websocket error, the bhyve websocket portion of the log from the reboot is:

```
2020-07-03 06:32:19 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Authenticating websocket
2020-07-03 06:32:19 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket connected
2020-07-03 06:33:39 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.547296289232] Client unable to keep up with pending messages. Stayed over 512 for 5 seconds
2020-07-03 06:33:45 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.547296289232] Client unable to keep up with pending messages. Stayed over 512 for 5 seconds
```

I am not certain if the pending messages comes from the bhyve websocket or not,

**Comment by sebr** [comment]
@txwindsurfer I have heard of other reports that when the websocket is disconnected unexpectedly (eg a network connection drops out) that the integration doesn't correctly handle this case. I'll take a look when I can.

The error logs from the `homeassistant.components.websocket_api.http.connection` component are not related - these are from the websocket API which HA uses for lovelace frontend updates.

**Comment by sebr** [comment]
I just noticed this happen, will look into it more. Here are my logs:

```
2020-07-15 06:12:09 DEBUG (MainThread) [custom_components.bhyve.pybhyve.websocket] msg received WSMessage(type=<WSMsgType.CLOSED: 257>, data=None, extra=None)
2020-07-15 06:12:09 ERROR (MainThread) [custom_components.bhyve.pybhyve.websocket] websocket connection closed
2020-07-15 06:12:09 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket closed? True
2020-07-15 06:12:09 WARNING (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket exception: 
2020-07-15 06:12:09 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Reconnecting websocket; state: running
2020-07-15 06:12:09 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Reconnecting to Orbit in 5.
2020-07-15 06:12:25 ERROR (MainThread) [custom_components.bhyve.pybhyve.websocket] Unexpected error 502, message='Invalid response status', url='wss://api.orbitbhyve.com/v1/events
```

**Comment by txwindsurfer** [comment]
I am glad you saw something happen.  My bhyve didn't water this morning and no HA or .bhyve log error was produced.  The error you show would have happened between my waterings so that is likely the cause of my failure.  Turning on the switch didn't produce an error either when the websock was down and of course bhyve didn't receive the signal.  I changed my bhyve logging to info from debug so maybe I will catch it next time.  I rebooted HA and bhyve works again.  Thanks so much for looking into it!

**Comment by sebr** [comment]
Yep that's consistent with what I saw. Just need to be able to reliably reproduce.

**Comment by txwindsurfer** [comment]
Happened again July 21 at 2:37pm CDT.  My HA logs show the following errors/warnings:
First:
```
Log Details (ERROR)
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:116
Integration: Orbit BHyve Integration (documentation)
First occurred: July 21, 2020, 2:37:32 PM (1 occurrences)
Last logged: July 21, 2020, 2:37:32 PM

websocket connection closed

```
Second:
```
Log Details (WARNING)
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:127
Integration: Orbit BHyve Integration (documentation)
First occurred: July 21, 2020, 2:37:32 PM (1 occurrences)
Last logged: July 21, 2020, 2:37:32 PM

Websocket exception:
```
And third:
```
Log Details (ERROR)
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:130
Integration: Orbit BHyve Integration (documentation)
First occurred: July 21, 2020, 2:37:37 PM (1 occurrences)
Last logged: July 21, 2020, 2:37:37 PM

Client connection error

```
BHyve is nonresponsive but shows available.  The error on 7/15 showed my Bhyve controller as unavailable.  This time the devices show as available but Bhyve controller is nonresponsive.

There a no entries in the .bhyve/packets.dmp.  If I try to use the switch for a zone, no entries appear in the .dmp which appears to have also closed.

So socket was lost and HA shows devices are available but there is  no response from my Bhyve controller.  No entries are now appearing in the log.

Rebooted HA and all is working again.

Thanks for your efforts on this integration!  Let me know if I can do or provide additional information or debugging help.

**Comment by txwindsurfer** [comment]
I don't know if this helps but I had 2 entries in the HA log (Error and Warning shown below) on 7/29 but the Bhyve controller still worked without an HA reboot.  The difference between the event on 7/15 when the controller connection was lost and required an HA reboot and the 7/29 event when the controller did not need an HA reboot was the websocket.py:130 entry occured when a reboot was required.

Log Details (ERROR)
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:116
Integration: Orbit BHyve Integration (documentation)
First occurred: July 29, 2020, 1:31:49 PM (1 occurrences)
Last logged: July 29, 2020, 1:31:49 PM

websocket connection closed


Log Details (WARNING)
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:127
Integration: Orbit BHyve Integration (documentation)
First occurred: July 29, 2020, 1:31:49 PM (1 occurrences)
Last logged: July 29, 2020, 1:31:49 PM

Websocket exception:

Full log entry:

`2020-07-29 13:31:49 DEBUG (MainThread) [custom_components.bhyve.pybhyve.websocket] msg received WSMessage(type=<WSMsgType.CLOSED: 257>, data=None, extra=None)
2020-07-29 13:31:49 ERROR (MainThread) [custom_components.bhyve.pybhyve.websocket] websocket connection closed
2020-07-29 13:31:49 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket closed? True
2020-07-29 13:31:49 WARNING (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket exception: 
2020-07-29 13:31:49 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Reconnecting websocket; state: running
2020-07-29 13:31:49 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Reconnecting to Orbit in 5.
2020-07-29 13:32:07 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Authenticating websocket
2020-07-29 13:32:07 INFO (MainThread) [custom_components.bhyve.pybhyve.websocket] Websocket connected
`

**Comment by sebr** [comment]
Thanks @txwindsurfer - most of the time the when the websocket connection is closed it reconnects - this is what you're seeing in your latest logs, and generally good news 😄.

I'll release a new version with some better/clearer logging.

**Comment by sebr** [comment]
I have been testing some improvements over the past few weeks and the integration is now more resilient to connectivity drops. I've released v`2.0.3` with these fixes so I'll close this issue now. Please reopen if you see any other problems!

**Comment by txwindsurfer** [comment]
@sebr:

My Bhyve connection closed today and was unable to reconnect without a restart of HA.  I received these errors and the packets.dump did not show any entries during this time.  It seemed to be restarting successfully for me until this last update.
Thanks for your work on this component!


Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:117
Integration: Orbit BHyve Integration (documentation)
First occurred: 10:36:34 AM (2 occurrences)
Last logged: 11:05:13 AM

websocket connection closed

followed by:
Logger: custom_components.bhyve.pybhyve.websocket
Source: custom_components/bhyve/pybhyve/websocket.py:137
Integration: Orbit BHyve Integration (documentation)
First occurred: 11:05:49 AM (1 occurrences)
Last logged: 11:05:49 AM

Client connection error; state: starting
